File: src/lira/linguistics/role/clause_reader.py
"""ClauseReader: spec 13.3's actual `Clause.read()` implementation.
Reads a token span as a sequence of Phrases (via PhraseReader), then
assigns each phrase a clause role (subject/predicate/object/complement/
modifier) against GrammarConfigurator.clause_element_templates. Only
ClauseType.INDEPENDENT has a populated template in this phase -- see
clause_type.py -- so this always attempts exactly one flat, non-
recursive independent-clause reading over its whole span; relative/
dependent/coordinated clauses (clause-level recursion) are Phase 2.

Clause validity is not simply the worst of its phrases' validity: spec
20's own worked example ("The fox over the dog.") has two individually
VALID phrases (a NOUN_PHRASE, a PREPOSITIONAL_PHRASE) but an INVALID
clause, because no VERB_PHRASE predicate exists at all. So this combines
its own template-level check (subject/predicate/finite-verb presence)
with the worst outcome among its phrases -- see _validate."""

from typing import List, Optional, Sequence, Tuple

from ..data.clause import Clause
from ..data.clause_type import ClauseType
from ..data.linguistic_unit_kind import LinguisticUnitKind
from ..data.phrase import Phrase
from ..data.reading_error import ReadingError, ReadingErrorKind
from ..data.token_reading import TokenReading
from ..data.validation_outcome import ValidationOutcome
from .grammar_configurator import ClauseTemplate, GrammarConfigurator
from .phrase_reader import PhraseReader
from .reading_scorer import ScoringFactors
from .sequence_engine import SequenceEngine

# Copular/linking-verb forms -- when the predicate's own head word is
# one of these, a following NOUN_PHRASE/ADJECTIVE_PHRASE is a complement
# ("A meaning IS a representation."), not an object; anything else with
# a following NOUN_PHRASE reads that phrase as an object instead. A
# closed, hand-picked set rather than a morphological test, since real
# transitivity/linking-verb classification isn't seeded data this phase
# has access to (see this file's module docstring on Phase 1 scope).
_LINKING_VERB_FORMS = {"is", "are", "was", "were", "be", "been", "being", "am"}


class ClauseReader:
    def __init__(self, phrase_reader: PhraseReader, engine: SequenceEngine, grammar: GrammarConfigurator):
        self.phrase_reader = phrase_reader
        self.engine = engine
        self.grammar = grammar

    def read(
        self,
        tokens: Sequence[TokenReading],
        *,
        start_index: int = 0,
        end_index: Optional[int] = None,
        grammar: Optional[GrammarConfigurator] = None,
    ) -> Clause:
        active_grammar = grammar or self.grammar
        end_index = len(tokens) if end_index is None else end_index
        template = active_grammar.clause_element_templates.get(ClauseType.INDEPENDENT)

        if template is None or start_index >= end_index:
            return self._empty_clause(tokens, start_index, end_index)

        phrases: List[Phrase] = []
        index = start_index
        while index < end_index:
            token = tokens[index]
            if token.is_punctuation:
                index += 1
                continue
            phrase = self.phrase_reader.read(tokens, start_index=index, end_index=end_index, grammar=active_grammar)
            phrases.append(phrase)
            index = phrase.end_position if phrase.end_position > index else index + 1

        subject, predicate, obj, complement, modifiers = self._assign_roles(phrases, template)
        finite_verb = predicate.head_word if predicate is not None else None

        validation, own_errors = self._validate(phrases, subject, predicate, template)
        phrase_errors = tuple(error for phrase in phrases for error in self._all_phrase_errors(phrase))
        errors = tuple(own_errors) + phrase_errors

        factors = ScoringFactors(
            validation=validation,
            unresolved_token_count=sum(1 for phrase in phrases if phrase.validation == ValidationOutcome.UNRESOLVED),
            undischarged_obligation_count=sum(len(phrase.open_obligations) for phrase in phrases),
            finite_verb_phrase_count=1 if predicate is not None else 0,
            phrase_count=len(phrases),
            lexical_evidence_sum=sum(phrase.confidence for phrase in phrases),
        )
        confidence = self.engine.scorer.confidence(factors)

        all_words = [word for phrase in phrases for word in phrase.words]
        clause = Clause(
            text=" ".join(word.text for word in all_words),
            tokens=all_words,
            is_independent=True,
            clause_type=ClauseType.INDEPENDENT,
            phrases=phrases,
            subject=subject, predicate=predicate, object=obj, complement=complement, modifiers=modifiers,
            finite_verb=finite_verb,
            start_position=start_index, end_position=index,
            validation=validation, confidence=confidence, errors=errors,
        )
        # Phase 1 reads at most one clause per sentence (clause_type.py),
        # so 0 is always this clause's own sequence number within its
        # sentence -- see phrase_reader.py's own note on the same point.
        clause.system_property = self.phrase_reader.graph_processor.create_property_wrapper(
            clause, LinguisticUnitKind.Clause, 0, "ClauseReader_ReadLayer",
        )
        return clause

    def _assign_roles(
        self, phrases: Sequence[Phrase], template: ClauseTemplate,
    ) -> Tuple[Optional[Phrase], Optional[Phrase], Optional[Phrase], Optional[Phrase], List[Phrase]]:
        subject: Optional[Phrase] = None
        predicate: Optional[Phrase] = None
        obj: Optional[Phrase] = None
        complement: Optional[Phrase] = None
        modifiers: List[Phrase] = []

        for phrase in phrases:
            if phrase.phrase_type is None:
                continue
            if predicate is None:
                if subject is None and phrase.phrase_type in template.subject_phrase_types:
                    subject = phrase
                    continue
                if phrase.phrase_type in template.predicate_phrase_types:
                    predicate = phrase
                    continue
                modifiers.append(phrase)
                continue

            # Past the predicate: first NOUN_PHRASE/ADJECTIVE_PHRASE
            # becomes the object or complement (never both), everything
            # else is a modifier.
            if obj is None and complement is None and phrase.phrase_type in template.complement_phrase_types:
                is_linking = predicate.head_word is not None and predicate.head_word.text.lower() in _LINKING_VERB_FORMS
                if not is_linking and phrase.phrase_type in template.object_phrase_types:
                    obj = phrase
                else:
                    complement = phrase
                continue
            modifiers.append(phrase)

        return subject, predicate, obj, complement, modifiers

    def _validate(
        self, phrases: Sequence[Phrase], subject: Optional[Phrase], predicate: Optional[Phrase], template: ClauseTemplate,
    ) -> Tuple[ValidationOutcome, List[ReadingError]]:
        errors: List[ReadingError] = []
        own_outcome = ValidationOutcome.VALID

        if template.subject_required and subject is None:
            own_outcome = ValidationOutcome.INVALID
            errors.append(ReadingError(
                kind=ReadingErrorKind.NO_VALID_CLAUSE_SEQUENCE, level=LinguisticUnitKind.Clause,
                message="No subject-shaped phrase found before the predicate",
            ))
        if template.predicate_required and predicate is None:
            own_outcome = ValidationOutcome.INVALID
            errors.append(ReadingError(
                kind=ReadingErrorKind.MISSING_PREDICATE, level=LinguisticUnitKind.Clause,
                message="No VERB_PHRASE found for this clause's predicate",
            ))
        elif predicate is not None and predicate.head_part_of_speech not in template.predicate_head_requires:
            own_outcome = ValidationOutcome.INVALID
            errors.append(ReadingError(
                kind=ReadingErrorKind.MISSING_FINITE_VERB, level=LinguisticUnitKind.Clause,
                message="The clause's predicate has no finite verb form",
            ))

        phrase_outcomes = [phrase.validation for phrase in phrases if phrase.phrase_type is not None]
        worst = min([own_outcome] + phrase_outcomes, key=lambda outcome: outcome.value) if phrase_outcomes else own_outcome
        return worst, errors

    def _all_phrase_errors(self, phrase: Phrase) -> List[ReadingError]:
        """A phrase's own `.errors` deliberately excludes its nested
        phrases' errors (phrase_reader.py's own _build_errors docstring:
        each nested Phrase, e.g. a PREPOSITIONAL_PHRASE's object, carries
        its own). Clause/Sentence-level error reporting (spec 21) still
        needs every error reachable from one place, so this walks
        nested_phrases to collect them -- an unknown word inside a PP's
        object (e.g. "over the dog" with "dog" unseeded) must surface at
        the sentence level, not only on that inner NOUN_PHRASE object."""
        errors = list(phrase.errors)
        for nested in phrase.nested_phrases:
            errors.extend(self._all_phrase_errors(nested))
        return errors

    def _empty_clause(self, tokens: Sequence[TokenReading], start_index: int, end_index: int) -> Clause:
        error = ReadingError(
            kind=ReadingErrorKind.NO_VALID_CLAUSE_SEQUENCE, level=LinguisticUnitKind.Clause,
            message="Empty token span or no clause template configured for ClauseType.INDEPENDENT",
            token_index=start_index if start_index < len(tokens) else None,
        )
        clause = Clause(
            text="", tokens=[], is_independent=None, clause_type=None,
            start_position=start_index, end_position=end_index,
            validation=ValidationOutcome.INVALID, confidence=0.0, errors=(error,),
        )
        clause.system_property = self.phrase_reader.graph_processor.create_property_wrapper(
            clause, LinguisticUnitKind.Clause, 0, "ClauseReader_ReadLayer",
        )
        return clause

File: prototype/src/lira/linguistics/role/clause_reader.ts
import { createClause, type Clause } from "../data/clause";
import { ClauseType } from "../data/clause_type";
import { LinguisticUnitKind } from "../data/linguistic_unit_kind";
import { createMainClause } from "../data/main_clause";
import { createSubordinateClause, type SubordinateClauseType } from "../data/subordinate_clause";
import type { Phrase } from "../data/phrase";
import type { Word } from "../../vocabulary/data/entities/word";
import { createReadingError, ReadingErrorKind, type ReadingError } from "../data/reading_error";
import { isPunctuation, type TokenReading } from "../data/token_reading";
import { ValidationOutcome } from "../data/validation_outcome";
import { embeddedSubjectClauseSpan, embeddingTrigger } from "./clause_embedding";
import type { ClauseTemplate, GrammarConfigurator } from "./grammar_configurator";
import type { PhraseReader } from "./phrase_reader";
import { createScoringFactors } from "./reading_scorer";
import type { SequenceEngine } from "./sequence_engine";

/** ClauseReader: spec 13.3's actual readClause() implementation. Reads
 * a token span as a sequence of Phrases (via PhraseReader), then
 * assigns each phrase a clause role (subject/predicate/object/
 * complement/modifier) against GrammarConfigurator.clauseElementTemplates.
 * Only ClauseType.INDEPENDENT has a populated template in this phase --
 * see clause_type.ts -- so this always attempts exactly one flat,
 * non-recursive independent-clause reading over its whole span;
 * relative/dependent/coordinated clauses (clause-level recursion) are
 * Phase 2.
 *
 * Clause validity is not simply the worst of its phrases' validity:
 * spec 20's own worked example ("The fox over the dog.") has two
 * individually VALID phrases (a NOUN_PHRASE, a PREPOSITIONAL_PHRASE)
 * but an INVALID clause, because no VERB_PHRASE predicate exists at
 * all. So this combines its own template-level check (subject/
 * predicate/finite-verb presence) with the worst outcome among its
 * phrases -- see validate.
 *
 * Ported from linguistics/role/clause_reader.py. */

// Copular/linking-verb forms -- when the predicate's own head word is
// one of these, a following NOUN_PHRASE/ADJECTIVE_PHRASE is a
// complement ("A meaning IS a representation."), not an object;
// anything else with a following NOUN_PHRASE reads that phrase as an
// object instead. A closed, hand-picked set rather than a
// morphological test, since real transitivity/linking-verb
// classification isn't seeded data this phase has access to.
const LINKING_VERB_FORMS = new Set(["is", "are", "was", "were", "be", "been", "being", "am"]);

export interface ClauseReadOptions {
  startIndex?: number;
  endIndex?: number;
  grammar?: GrammarConfigurator;
  trace?: unknown[];
  // Which ClauseTemplate this read attempts -- INDEPENDENT (the default,
  // every real top-level read) or DEPENDENT, the one other populated
  // template (clause_embedding.ts's own docstring): ClauseReader.read()
  // recurses into itself with this set to DEPENDENT for a candidate
  // embedded-subject-clause span, never called this way from outside
  // this file. RELATIVE/COORDINATED still have no template at all
  // (grammar_configurator.ts), so passing either here still reports
  // UNRESOLVED via emptyClause() below, unchanged.
  clauseType?: ClauseType;
  // Whether this read attempts its own embedded-subject-clause
  // recognition -- true by default, false on the one recursive call
  // read() itself makes to try a candidate embedded span. Required,
  // not incidental: a free-relative trigger ("what") IS the embedded
  // clause's own subject (embeddingTrigger()'s own docstring,
  // clause_embedding.ts), so the recursive read's span starts at that
  // exact same token -- without this guard it would immediately
  // recognise "what" as a trigger all over again and recurse forever.
  // The same "one level only" scoping this session's own Coordination
  // work already settled on for a coordinate side
  // (resolveCoordinateSide(), vocabulary/role/processor/phrase_processor.ts) --
  // an embedded clause is never itself searched for a second, nested
  // embedded clause.
  allowEmbedding?: boolean;
}

export class ClauseReader {
  constructor(
    public readonly phraseReader: PhraseReader,
    private readonly engine: SequenceEngine,
    private readonly grammar: GrammarConfigurator,
  ) {}

  /** `trace`, when an array is passed, is threaded straight through to
   * every PhraseReader.read() call this clause makes. Purely additive/
   * observational. */
  read(tokens: readonly TokenReading[], options: ClauseReadOptions = {}): Clause {
    const activeGrammar = options.grammar ?? this.grammar;
    const startIndex = options.startIndex ?? 0;
    const endIndex = options.endIndex ?? tokens.length;
    const clauseType = options.clauseType ?? ClauseType.INDEPENDENT;
    const template = activeGrammar.clauseElementTemplates.get(clauseType);

    if (template === undefined || startIndex >= endIndex) {
      return this.emptyClause(tokens, startIndex, endIndex);
    }

    const phrases: Phrase[] = [];
    const orderedWordChunks: Word[][] = [];
    let resolvedSubject: Phrase | Clause | undefined;
    const nestedClauses: Clause[] = [];
    let index = startIndex;
    while (index < endIndex) {
      const token = tokens[index];
      if (isPunctuation(token)) {
        index += 1;
        continue;
      }
      // Embedded-subject-clause recognition (clause_embedding.ts's own
      // docstring) -- only attempted once, and only before a subject-
      // shaped phrase has already been read: "no subject found yet" is
      // a purely local check over `phrases` so far, the same
      // subjectPhraseTypes set assignRoles() below itself checks, not a
      // duplicate of that function's own full role-resolution state
      // machine.
      if (
        (options.allowEmbedding ?? true) &&
        resolvedSubject === undefined &&
        !phrases.some((phrase) => phrase.phraseType !== undefined && template.subjectPhraseTypes.has(phrase.phraseType))
      ) {
        const embeddedStart = embeddingTrigger(token);
        if (embeddedStart !== undefined) {
          const found = embeddedSubjectClauseSpan(embeddedStart, endIndex, tokens, this.engine, template, (spanStart, boundary) =>
            this.read(tokens, { startIndex: spanStart, endIndex: boundary, grammar: activeGrammar, clauseType: ClauseType.DEPENDENT, allowEmbedding: false, trace: options.trace }));
          if (found !== undefined) {
            resolvedSubject = found.embedded;
            nestedClauses.push(found.embedded);
            // A complementizer trigger ("that") sits at `index`, one
            // position *before* `embeddedStart` -- consumed here, never
            // becoming a Phrase of its own, but still a real word this
            // clause's own text contains. Materialise it directly
            // (never routed through phraseReader.read(), the only other
            // place a token normally becomes a Word) so it's never
            // silently absent from `clause.text`/`.tokens` the way a
            // dropped token once made "is" vanish from a read Sentence
            // entirely (this log's own AUXILIARY/`is` mystery section) --
            // a free-relative trigger ("what") needs no such step, it's
            // already the embedded clause's own first token, included in
            // `found.embedded.tokens` below.
            if (embeddedStart !== index) {
              const triggerWord = this.phraseReader.graphProcessor.materialiseToken(token, token.tokenIndex, token.candidates[0]);
              orderedWordChunks.push([triggerWord]);
            }
            orderedWordChunks.push(found.embedded.tokens);
            index = found.boundary;
            continue;
          }
        }
      }
      const phrase = this.phraseReader.read(tokens, { startIndex: index, endIndex, grammar: activeGrammar, trace: options.trace });
      phrases.push(phrase);
      orderedWordChunks.push(phrase.words);
      index = phrase.endPosition > index ? phrase.endPosition : index + 1;
    }

    const { subject, predicate, obj, complement, modifiers } = this.assignRoles(phrases, template, resolvedSubject);
    const finiteVerb = predicate?.headWord;

    const { validation, errors: ownErrors } = this.validate(phrases, subject, predicate, template);
    const phraseErrors = phrases.flatMap((phrase) => this.allPhraseErrors(phrase));
    const nestedClauseErrors = nestedClauses.flatMap((nested) => nested.errors);
    const errors: readonly ReadingError[] = [...ownErrors, ...phraseErrors, ...nestedClauseErrors];

    const factors = createScoringFactors({
      validation,
      unresolvedTokenCount: phrases.filter((phrase) => phrase.validation === ValidationOutcome.UNRESOLVED).length,
      undischargedObligationCount: phrases.reduce((sum, phrase) => sum + phrase.openObligations.length, 0),
      finiteVerbPhraseCount: predicate !== undefined ? 1 : 0,
      phraseCount: phrases.length,
      lexicalEvidenceSum: phrases.reduce((sum, phrase) => sum + phrase.confidence, 0),
    });
    const confidence = this.engine.scorer.confidence(factors);

    const allWords = orderedWordChunks.flat();
    const clauseInit = {
      text: allWords.map((word) => word.text).join(" "),
      tokens: allWords,
      phrases,
      subject, predicate, object: obj, complement, modifiers,
      finiteVerb, nestedClauses,
      startPosition: startIndex, endPosition: index,
      validation, confidence, errors,
    };
    // INDEPENDENT (the one ClauseType every real top-level read still
    // resolves to) builds a MainClause; DEPENDENT (this file's own
    // recursive embedded-subject-clause read, above) builds a real
    // SubordinateClause -- the first ClauseType.DEPENDENT read/RELATIVE/
    // COORDINATED still have no template at all (grammar_configurator.ts),
    // so those still can't reach this line at all, only emptyClause()
    // above.
    const clause = clauseType === ClauseType.INDEPENDENT
      ? createMainClause(clauseInit)
      : createSubordinateClause({ ...clauseInit, clauseType: clauseType as SubordinateClauseType });
    // Phase 1 reads at most one clause per sentence (clause_type.ts),
    // so 0 is always this clause's own sequence number within its
    // sentence.
    clause.systemProperty = this.phraseReader.graphProcessor.createPropertyWrapper(
      clause, LinguisticUnitKind.Clause, 0, "ClauseReader_ReadLayer",
    );
    return clause;
  }

  private assignRoles(
    phrases: readonly Phrase[], template: ClauseTemplate, resolvedSubject?: Phrase | Clause,
  ): { subject?: Phrase | Clause; predicate?: Phrase; obj?: Phrase; complement?: Phrase; modifiers: Phrase[] } {
    let subject: Phrase | Clause | undefined = resolvedSubject;
    let predicate: Phrase | undefined;
    let obj: Phrase | undefined;
    let complement: Phrase | undefined;
    const modifiers: Phrase[] = [];

    for (const phrase of phrases) {
      if (phrase.phraseType === undefined) continue;
      if (predicate === undefined) {
        if (subject === undefined && template.subjectPhraseTypes.has(phrase.phraseType)) {
          subject = phrase;
          continue;
        }
        if (template.predicatePhraseTypes.has(phrase.phraseType)) {
          predicate = phrase;
          continue;
        }
        modifiers.push(phrase);
        continue;
      }

      // Past the predicate, no subject found yet -- subject-auxiliary
      // inversion ("Did the young woman open the gate?": the AUXILIARY
      // "Did" alone already satisfies predicateHeadRequires on its own,
      // long before the real main verb "open" is even reached, so the
      // ordinary pre-predicate scan above never finds a subject at
      // all). The first subject-eligible phrase found now is the
      // (inverted) subject, not the object -- real English word order
      // for a yes/no question, a fronted wh-question, or "Is behind the
      // station a safe place to wait?"'s own PrepositionalPhrase
      // subject alike. Never fires for an ordinary declarative, where
      // the pre-predicate scan above already found the subject.
      if (subject === undefined && template.subjectPhraseTypes.has(phrase.phraseType)) {
        subject = phrase;
        continue;
      }

      // Past the predicate: first NOUN_PHRASE/ADJECTIVE_PHRASE becomes
      // the object or complement (never both), everything else is a
      // modifier.
      if (obj === undefined && complement === undefined && template.complementPhraseTypes.has(phrase.phraseType)) {
        const isLinking = predicate.headWord !== undefined && LINKING_VERB_FORMS.has(predicate.headWord.text.toLowerCase());
        if (!isLinking && template.objectPhraseTypes.has(phrase.phraseType)) {
          obj = phrase;
        } else {
          complement = phrase;
        }
        continue;
      }
      modifiers.push(phrase);
    }

    return { subject, predicate, obj, complement, modifiers };
  }

  private validate(
    phrases: readonly Phrase[], subject: Phrase | Clause | undefined, predicate: Phrase | undefined, template: ClauseTemplate,
  ): { validation: ValidationOutcome; errors: ReadingError[] } {
    const errors: ReadingError[] = [];
    let ownOutcome = ValidationOutcome.VALID;

    if (template.subjectRequired && subject === undefined) {
      ownOutcome = ValidationOutcome.INVALID;
      errors.push(createReadingError({
        kind: ReadingErrorKind.NO_VALID_CLAUSE_SEQUENCE, level: LinguisticUnitKind.Clause,
        message: "No subject-shaped phrase found before the predicate",
      }));
    }
    if (template.predicateRequired && predicate === undefined) {
      ownOutcome = ValidationOutcome.INVALID;
      errors.push(createReadingError({
        kind: ReadingErrorKind.MISSING_PREDICATE, level: LinguisticUnitKind.Clause,
        message: "No VERB_PHRASE found for this clause's predicate",
      }));
    } else if (predicate !== undefined && (predicate.headPartOfSpeech === undefined || !template.predicateHeadRequires.has(predicate.headPartOfSpeech))) {
      ownOutcome = ValidationOutcome.INVALID;
      errors.push(createReadingError({
        kind: ReadingErrorKind.MISSING_FINITE_VERB, level: LinguisticUnitKind.Clause,
        message: "The clause's predicate has no finite verb form",
      }));
    }

    const phraseOutcomes = phrases.filter((phrase) => phrase.phraseType !== undefined).map((phrase) => phrase.validation);
    // `subject` is only ever a Clause when embeddedSubjectClauseSpan()
    // (clause_embedding.ts) already required it to be VALID before
    // accepting that boundary -- included here anyway rather than relied
    // on silently, so this reduction stays correct even if that
    // invariant ever changes.
    const outcomes = subject !== undefined && "clauseType" in subject ? [ownOutcome, subject.validation, ...phraseOutcomes] : [ownOutcome, ...phraseOutcomes];
    const worst = outcomes.reduce((worstSoFar, outcome) => (outcome < worstSoFar ? outcome : worstSoFar));
    return { validation: worst, errors };
  }

  /** A phrase's own `.errors` deliberately excludes its nested
   * phrases' errors, so Clause/Sentence-level error reporting (spec
   * 21) still needs every error reachable from one place -- this walks
   * nestedPhrases to collect them. */
  private allPhraseErrors(phrase: Phrase): ReadingError[] {
    const errors = [...phrase.errors];
    for (const nested of phrase.nestedPhrases) errors.push(...this.allPhraseErrors(nested));
    return errors;
  }

  private emptyClause(tokens: readonly TokenReading[], startIndex: number, endIndex: number): Clause {
    const error = createReadingError({
      kind: ReadingErrorKind.NO_VALID_CLAUSE_SEQUENCE, level: LinguisticUnitKind.Clause,
      message: "Empty token span or no clause template configured for ClauseType.INDEPENDENT",
      tokenIndex: startIndex < tokens.length ? startIndex : undefined,
    });
    const clause = createClause({
      text: "", tokens: [], isIndependent: undefined, clauseType: undefined,
      startPosition: startIndex, endPosition: endIndex,
      validation: ValidationOutcome.INVALID, confidence: 0.0, errors: [error],
    });
    clause.systemProperty = this.phraseReader.graphProcessor.createPropertyWrapper(
      clause, LinguisticUnitKind.Clause, 0, "ClauseReader_ReadLayer",
    );
    return clause;
  }
}
